lib/id3c/api/datastore.py

```diff
@@ -382,6 +383,63 @@ def fetch_identifier_set_uses(session: DatabaseSession) -> Any:
 
         return list(cursor)
 
+@export
+@catch_permission_denied
+def store_sample(session: DatabaseSession, sample: dict) -> Any:
+    """"
+    Validate the given *sample* and insert or update in the backing database.
+
+    Returns a list of in the same order as the input, with each object including
+    the ``sample_id`` (string), ``status`` (string) to indicate if inserted,
+    updated, or validation failed, and ``details`` to indicate reason for
+    failed validation.
+    """
+    with session:
+        sample_barcode = sample.pop("sample_id", None)
+        sample_identifier = find_identifier(session, sample_barcode) if sample_barcode else None
+        collection_barcode = sample.pop("collection_id", None) 
+        collection_identifier = find_identifier(session, collection_barcode) if collection_barcode else None
+        
+        result = {
+            "sample_barcode": sample_barcode,
+            "collection_barcode": collection_barcode
+        }
+
+        # validate barcodes
+        if sample_barcode and not sample_identifier:
+            result["status"] = "validation_failed"
+            result["details"] = f"sample barcode «{sample_barcode}» not found"
+        elif sample_identifier and sample_identifier.set_use != 'sample':
+            result["status"] = "validation_failed"
+            result["details"] = f"barcode «{sample_barcode}» has use type «{sample_identifier.set_use}»"
+        elif collection_barcode and not collection_identifier:
+            result["status"] = "validation_failed"
+            result["details"] = f"collection barcode «{collection_barcode}» not found"
+        elif collection_identifier and collection_identifier.set_use != 'collection':
```

Comment:
It looks like you dropped the validation of what collection identifier sets we expect here. Was that intentional to allow samples with any collection identifier set to be updated with this API?

Reply:
Yes, the idea here was to utilize the relatively new `use` column in the `identifier_set` table to avoid having to maintain a hardcoded list of `expected_identifier_sets` that needs to be updated every time a new set is added, as is being done in the manifest ETL function (it can also be replaced in that context but I didn't want to make any changes there that are unrelated to this branch)

Reply:
Makes sense! In inspecting more closely, it looks like the list in manifest.py does match perfectly with the identifier sets with use `collection`. I was thinking there might be additional filtering going on to identifier sets we expected to see come through the manifest, but since not, this makes a whole lot of sense. And will be one less piece to maintain. Sounds good to me!